File: fonte/u00 Nivelamento/c/recursividade/contarmaiuscula.c
#include <stdio.h> 
#include <stdlib.h> 
#include <string.h>

#define bool int
#define true 1
#define false 0

bool isUpper(char x){
   return (x >= 'A' && x <= 'Z');
}

int contMaiusculo(char* s){
   int cont = 0;

   for(int i = 0; i < strlen(s); i++){
      if(isUpper(s[i]) == true){
         cont++;
      }
   }
   return cont;
}

int contMaiusculoRec(char* s, int i){
   int cont = 0;

   if(i < strlen(s)){
      if(isUpper(s[i]) == true){
         cont++;
      }
      cont += contMaiusculoRec(s, i + 1);
   }
   return cont;
}

int contMaiusculoRecursivo(char* s){
   return contMaiusculoRec(s, 0);
}

int main(int argc, char *argv[]) { 
   printf("\nAlGoritmos e Estrutudas de Dados: %i", contMaiusculo("AlGoritmos e Estruturas de Dados"));
   printf("\nAlGoritmos e Estruturas de Dados: %i", contMaiusculoRecursivo("AlGoritmos e Estruturas de Dados"));

   return 0; 
}

File: fonte/u00 Nivelamento/java/recursividade/ContarMaiuscula.java
/**
 * Recursividade
 * @author Max do Val Machado
 * @version 2 01/2015
 */
class ContarMaiuscula {
   public static boolean isUpper(char x){
      return (x >= 'A' && x <= 'Z');
   }

   public static int contMaiusculo(String s){
      int cont = 0;

      for(int i = 0; i < s.length(); i++){
         if(isUpper(s.charAt(i)) == true){
            cont++;
         }
      }
      return cont;
   }
   public static int contMaiusculo2(String s){
      return contMaiusculo2(s, 0);
   }
   public static int contMaiusculo2(String s, int i){
      int cont = 0;

      if(i < s.length()){
         if(isUpper(s.charAt(i)) == true){
            cont++;
         }
         cont += contMaiusculo2(s, i + 1);
      }
      return cont;
   }

   public static void main (String[] args){
      System.out.println("AlGoritmos e Estruturas de Dados: " + contMaiusculo("AlGoritmos e Estruturas de Dados"));
      System.out.println("AlGoritmos e Estruturas de Dados: " + contMaiusculo2("AlGoritmos e Estruturas de Dados"));
   }
}
